Skip docs with Docvalues in NumericLeafComparator

LuXugang opened the pull request:
### Description

Could we implement `TermOrdValLeafComparator`'s same logic that using NumericDocValues to skip docs in `NumericLeafComparator` if we could not get a iterator by bkd?

LuXugang committed:
Skip document by docValues

LuXugang committed:
add test

LuXugang committed:
spotless

LuXugang commented:
```java
  public void test() throws IOException {
    final Directory dir = newDirectory();
    IndexWriterConfig config =
            new IndexWriterConfig()
                    // Make sure to use the default codec, otherwise some random points formats that have
                    // large values for maxPointsPerLeaf might not enable skipping with only 10k docs
                    .setCodec(TestUtil.getDefaultCodec());
    final IndexWriter writer = new IndexWriter(dir, config);
    final int numDocs = atLeast(10000);
    final int missValuesNumDocs = numDocs / 2;
    for (int i = 0; i < numDocs; ++i) {
      final Document doc = new Document();
      if (i <= missValuesNumDocs) { // missing value document
      } else {
        doc.add(new NumericDocValuesField("my_field", i));
        doc.add(new LongPoint("my_field", i));
      }
      writer.addDocument(doc);
    }
    final IndexReader reader = DirectoryReader.open(writer);
    writer.close();
    // single threaded so totalHits is deterministic
    IndexSearcher searcher =
            newSearcher(reader, random().nextBoolean(), random().nextBoolean(), false);
    final int numHits = 3;
    final int totalHitsThreshold = 3;

    { // test that optimization is run with NumericDocValues when missing value is NOT competitive
      final SortField sortField = new SortField("my_field", SortField.Type.LONG, true);
      sortField.setMissingValue(0L); // missing value is not competitive
      final Sort sort = new Sort(sortField);
      CollectorManager<TopFieldCollector, TopFieldDocs> manager =
              TopFieldCollector.createSharedManager(sort, numHits, null, totalHitsThreshold);
      TopDocs topDocs = searcher.search(new MatchAllDocsQuery(), manager);
      assertEquals(topDocs.scoreDocs.length, numHits);
      assertEquals(
              topDocs.totalHits.value,
              numDocs); // assert that all documents were collected => optimization was not run
    }

    reader.close();
    dir.close();
  }
```

This test shows we could not skip document by bkd,   but could use `NumericDocValues` to skip docs inthis RP.

LuXugang committed:
add test

LuXugang committed:
d

jpountz commented:
I'm not clear if this change is still correct when there is another sort field after the one that gets optimized. It seems like it could skip hits that are still needed.

jpountz reviewed (commented):
My intuition is that the way to fix it would be to allow the comparator to know whether it's the only one, and if so it could make `isMissingValueCompetitive` return false when the bottom value is equal to the missing value.

LuXugang commented:
> I'm not clear if this change is still correct when there is another sort field after the one that gets optimized

Thanks @jpountz. Oh, you are right, Sorry for missing this.

> My intuition is that the way to fix it would be to allow the comparator to know whether it's the only one

The most simply way is to add a default boolean method like `isOnlyComparator` in `LeafFieldComparator` and call this method in `LeafCollector`.  Is it a good way or worth for the comparator to know whether it's the only one?

jpountz commented:
We probably need to change `SortField#getComparator`. We introduced `enableSkipping` in the past, so that the first comparator in the chain could know it can dynamically prune, maybe it needs to become an enum, something like `Pruning.(NONE|GREATER_THAN|GREATER_THAN_OR_EQUAL_TO)`, and we'd only use `GREATER_THAN_OR_EQUAL_TO` when there is a single comparator?

LuXugang committed:
add Pruning

LuXugang committed:
add test

jpountz commented:
Thanks for adding the enum. In my view, we now need the two following changes:
 - `isMissingValueCompetitive()` should return false if the missing value is equal to the bottom value and the pruning is SKIP_MORE
 - The competitive iterator can better tune the min/max values. For instance currently, if the bottom value is 5 and the sort is ascending, we'll use a range on [MIN_VALUE, 5] to filter competitive hits. But with SKIP_MORE, we could now make it [MIN_VALUE, 4]. FYI we have `NumericUtils#(add|subtract)` to add/subtract on binary representations of numbers.

LuXugang committed:
isMissingValueCompetitive() should return false if the missing value is equal to the bottom value and the pruning is SKIP_MORE

LuXugang committed:
add test

LuXugang committed:
The competitive iterator can better tune the min/max values

LuXugang committed:
add test

LuXugang commented:
Thanks for providing excellent advice.

> isMissingValueCompetitive() should return false if the missing value is equal to the bottom value and the pruning is SKIP_MORE

 addressed in [4a72bf6](https://github.com/apache/lucene/pull/12405/commits/4a72bf6d8741ee779f16c3f403bd01cbbdf49bbe) and [94d560c](https://github.com/apache/lucene/pull/12405/commits/94d560c061c76c4c14732d4bff75a920b9f1e97b)

> The competitive iterator can better tune the min/max values. For instance currently, if the bottom value is 5 and the sort is ascending, we'll use a range on [MIN_VALUE, 5] to filter competitive hits. But with SKIP_MORE, we could now make it [MIN_VALUE, 4]. FYI we have NumericUtils#(add|subtract) to add/subtract on binary representations of numbers.

addressed in [4d50317](https://github.com/apache/lucene/pull/12405/commits/4d503171e5316a5f14597f57cb0626ed62759fba) and [416c333](https://github.com/apache/lucene/pull/12405/commits/416c33325b73348c8dd962bbed69d6f13f18b8df)

jpountz commented on lucene/core/src/java/org/apache/lucene/search/comparators/DoubleComparator.java:
```diff
@@ -61,8 +63,12 @@ public LeafFieldComparator getLeafComparator(LeafReaderContext context) throws I
     public DoubleLeafComparator(LeafReaderContext context) throws IOException {
       super(context);
+      deltaOne = new byte[Double.BYTES];
+      DoublePoint.encodeDimension(1d, deltaOne, 0);
```
This doesn't look correct to me. E.g. if the bottom value is 4 and the sort is ascending, this would mistakenly ignore 3.5. We need to go to the next double, which can be done by subtracting 1 to the encoded representation:
```suggestion
      deltaOne[deltaOne.length-1] = 1;
```

LuXugang replied:
addressed in [ef9f917](https://github.com/apache/lucene/pull/12405/commits/ef9f9173bbba04871dd489187cbdc26505d64418)

jpountz commented on lucene/core/src/java/org/apache/lucene/search/comparators/NumericComparator.java:
```diff
@@ -309,34 +323,100 @@ private void updateSkipInterval(boolean success) {
+     * reverse == false, missing value is non-competitive when it great than maxPackedValue
+     */
+    protected abstract boolean isMissingValueNotCompetitive(
+        byte[] minPackedValue, byte[] maxPackedValue);
```
I don't think we need this anymore? It should alreay be covered by the changes that you made to `isMissingValueCompetitive()`?

LuXugang replied:
addressed in [69c7259](https://github.com/apache/lucene/pull/12405/commits/69c72591d895813a0d045ee170b26a5de3e0e47e)

jpountz commented on lucene/core/src/java/org/apache/lucene/search/comparators/IntComparator.java:
```diff
@@ -98,19 +99,30 @@ public void copy(int slot, int doc) throws IOException {
+        IntPoint.encodeDimension(reverse ? bottom + 1 : bottom - 1, packedValue, 0);
+      } else {
+        IntPoint.encodeDimension(bottom, packedValue, 0);
+      }
```
Nit: I would prefer this +1/-1 logic to be in `NumericComparator` and make `encoreBottom` just encode the bottom value.

Also note that we need to handle overflows, e.g. `bottom` could be Integer.MIN_VALUE already, so subtracting 1 would underflow. When this happens, it means that there wouldn't be any competitive hits anymore.

LuXugang replied:
addressed in [ef9f917](https://github.com/apache/lucene/pull/12405/commits/ef9f9173bbba04871dd489187cbdc26505d64418)

jpountz commented on lucene/core/src/java/org/apache/lucene/search/Pruning.java:
```diff
@@ -0,0 +1,31 @@
+public enum Pruning {
+  /** Not allowed to skip documents. */
+  NONE,
+  /** Allowed to skip documents. */
```
Add a bit more docs about what/how docs may be skipped, e.g.

```suggestion
  /** Allowed to skip documents that compare strictly better than the top value, or strictly worse than the bottom value. */
```

jpountz commented on lucene/core/src/java/org/apache/lucene/search/Pruning.java:
```diff
@@ -0,0 +1,31 @@
+  /** Allowed to skip documents. */
+  SKIP,
+  /**
+   * Allowed to skip documents. {@link LeafFieldComparator} is the only comparator so that could do
```
Likewise here, something like: `Allowed to skip documents that compare strictly better than the top value, or worse than or equal to the bottom value.`

I would like to have a better name than `SKIP_MORE`, which doesn't tell much about what more may be skipped. I'm not completely satisfied with my initial proposal but I think it was still a bit better? Pruning.NONE, Pruning.GREATER_THAN (replaces SKIP) and Pruning.GREATER_THAN_OR_EQUAL_TO (replaces SKIP_MORE)?

LuXugang committed:
TopDocs#merge can never skip so we could use Pruning.NONE all the time here for simplicity

LuXugang commented on lucene/core/src/java/org/apache/lucene/search/TopDocs.java:
```diff
@@ -159,7 +159,12 @@ public MergeSortQueue(Sort sort, TopDocs[] shardHits, Comparator<ScoreDoc> tieBr
+                1,
+                compIDX == 0
+                    ? (sortFields.length > 1 ? Pruning.SKIP : Pruning.SKIP_MORE)
+                    : Pruning.NONE);
```
[b60f6fc](https://github.com/apache/lucene/pull/12405/commits/b60f6fc433c5cf3c2caa827092f3b93593c974b8)

LuXugang committed:
add more docs about Pruning

LuXugang committed:
remove isMissingValueNotCompetitive()

LuXugang committed:
better name for Pruning

LuXugang committed:
Nit: I would prefer this +1/-1 logic to be in NumericComparator and make encoreBottom just encode the bottom value.

jpountz commented on lucene/core/src/java/org/apache/lucene/search/comparators/NumericComparator.java:
```diff
@@ -309,38 +322,109 @@ private void updateSkipInterval(boolean success) {
+     * value is 5, we will use a range on [MIN_VALUE, 4].
+     */
+    private void encodeBottom() {
+      if (pruning == Pruning.GREATER_THAN_OR_EQUAL_TO && isBottomMinOrMax() == false) {
```
Can we avoid introducing this `isBottomMinOrMax()` method, and instead:
 - only subtract `deltaOne` if `bottom` is not full of zeroes
 - only add `deltaOne` if `bottom` is not full of `0xff`

LuXugang replied:
> only subtract deltaOne if bottom is not full of zeroes
only add deltaOne if bottom is not full of 0xff

It seems only work for `int` and `long`:

```java
        float maxFloat = Float.MAX_VALUE;
        float minFloat = Float.MIN_VALUE;
        byte[] maxFloatByte = new byte[Float.BYTES];
        byte[] minFloatByte = new byte[Float.BYTES];
        FloatPoint.encodeDimension(maxFloat, maxFloatByte, 0);
        FloatPoint.encodeDimension(minFloat, minFloatByte, 0);
```
result was maxFloatByte: [-1, 127, -1, -1] and minFloat: [-128, 0, 0, 1]

jpountz reviewed (commented):
I left a suggestion, but this looks good to me in general. Can you add a CHANGES entry?

jpountz committed:
Try to simplify the skipping logic.

jpountz commented:
@LuXugang I pushed a commit with the changes I had in mind, what do you think?

jpountz commented on lucene/core/src/test/org/apache/lucene/search/TestSortOptimization.java:
```diff
@@ -149,6 +149,11 @@ public void testLongSortOptimization() throws IOException {
+  /**
+   * test that if we skip docs only with {@link org.apache.lucene.index.NumericDocValues}, result
+   * must be equal to no optimization case
+   */
```
did you mean to add a test there?

LuXugang committed:
spotless

LuXugang commented on lucene/core/src/java/org/apache/lucene/util/NumericUtils.java:
```diff
@@ -132,6 +132,52 @@ public static void add(int bytesPerDim, int dim, byte[] a, byte[] b, byte[] resu
+   * the current value. This returns false if, and only if, the value is currently equal to the
+   * maximum possible value.
+   */
+  public static boolean nextUp(byte[] bytes) {
```
`nextup` and `nextdown` only work well for int/long

double/float case can't pass:
```java
  public void testNextUpDouble() {
    for (double i : new double[] {Double.MIN_VALUE, Double.MAX_VALUE}) {
      byte[] b = new byte[Long.BYTES];
      long a = NumericUtils.doubleToSortableLong(i);
      NumericUtils.longToSortableBytes(a, b, 0);
      assertFalse(NumericUtils.nextUp(b));
      assertEquals(i, NumericUtils.sortableLongToDouble(NumericUtils.sortableBytesToLong(b, 0)), 0.0);
    }
  }
```

Because `Double.MIN_VALUE`, `Double.MAX_VALUE`'s byte values are: [-128, 0, 0, 0, 0, 0, 0, 1] and [-1, -17, -1, -1, -1, -1, -1, -1]

jpountz replied:
Is is fine. This is because there is no double that is represented by eights zero bytes or eight 0xFF bytes, so nextUp/nextDown never overflow/underflow and nextUp/nextDown never return `false`. If you pass +Infinity, it will return NaN, but this is fine from the perspective of the comparator, it just needs a value that compares greater/less than a given value in its encoded form.

LuXugang replied:
Thank you for your explanation! @jpountz

LuXugang committed:
top values could still competitive when queue is not full

LuXugang commented on lucene/core/src/java/org/apache/lucene/search/comparators/NumericComparator.java:
```diff
@@ -329,87 +324,65 @@ private void updateSkipInterval(boolean success) {
+    private void encodeTop() {
+      if (reverse == false) {
+        encodeTop(minValueAsBytes);
+        if (pruning == Pruning.GREATER_THAN_OR_EQUAL_TO) {
```
top values could still competitive when queue is not full. instead: 
`if (pruning == Pruning.GREATER_THAN_OR_EQUAL_TO && queueFull)`

see this case:
```java
public void testSortOptimizationEqualValues1() throws IOException{
    final Directory dir = newDirectory();
    IndexWriterConfig config = new IndexWriterConfig().setCodec(TestUtil.getDefaultCodec());
    final IndexWriter writer = new IndexWriter(dir, config);
    final int numDocs = atLeast(TEST_NIGHTLY ? 50_000 : 10_000);
    for (int i = 1; i <= numDocs; ++i) {
      final Document doc = new Document();
      doc.add(new NumericDocValuesField("my_field1", 100)); // all docs have the same value of my_field1
      doc.add(new IntPoint("my_field1", 100));
      writer.addDocument(doc);
    }
    writer.flush();
    final IndexReader reader = DirectoryReader.open(writer);
    writer.close();
    IndexSearcher searcher = newSearcher(reader, random().nextBoolean(), random().nextBoolean(), false);
    final int numHits = 3;
    final int totalHitsThreshold = 3;
    final int afterValue = 100;
    final int afterDocID = 10 + random().nextInt(1000);
    final SortField sortField = new SortField("my_field1", SortField.Type.INT);
    final Sort sort = new Sort(sortField);
    FieldDoc after = new FieldDoc(afterDocID, Float.NaN, new Integer[] {afterValue});
    CollectorManager<TopFieldCollector, TopFieldDocs> manager = TopFieldCollector.createSharedManager(sort, numHits, after, totalHitsThreshold);
    TopDocs topDocs1 = searcher.search(new MatchAllDocsQuery(), manager);
    assertEquals(topDocs1.scoreDocs.length, numHits);

    // without optimization
    sortField.setOptimizeSortWithPoints(false);
    manager = TopFieldCollector.createSharedManager(sort, numHits, after, totalHitsThreshold);
    TopDocs topDocs2 = searcher.search(new MatchAllDocsQuery(), manager);
    assertEquals(topDocs1.scoreDocs.length, topDocs2.scoreDocs.length);
    reader.close();
    dir.close();
  }
```

jpountz replied:
Oh, indeed, good catch! Can you fix it?

LuXugang replied:
> Oh, indeed, good catch! Can you fix it?

addressed in [9d30e0b](https://github.com/apache/lucene/pull/12405/commits/9d30e0b71c0c8a787f352bc39b2faf0c025810fb)

LuXugang commented:
> @LuXugang I pushed a commit with the changes I had in mind, what do you think?

It is nice simplication. Thanks.

LuXugang commented:
There is still a bug in this RP, test failed with
`./gradlew test --tests TestSortOptimization.testRandomLong -Dtests.seed=6B2B316B7080952B -Dtests.locale=yav-Latn-CM -Dtests.timezone=Indian/Mahe -Dtests.asserts=true -Dtests.file.encoding=UTF-8`

I would fix it later.

LuXugang committed:
fix ./gradlew test --tests TestSortOptimization.testRandomLong -Dtests.seed=6B2B316B7080952B -Dtests.locale=yav-Latn-CM -Dtests.timezone=Indian/Mahe -Dtests.asserts=true -Dtests.file.encoding=UTF-8

LuXugang committed:
add test

LuXugang committed:
resolve conflict

LuXugang commented:
> There is still a bug in this RP, test failed with `./gradlew test --tests TestSortOptimization.testRandomLong -Dtests.seed=6B2B316B7080952B -Dtests.locale=yav-Latn-CM -Dtests.timezone=Indian/Mahe -Dtests.asserts=true -Dtests.file.encoding=UTF-8`
> 
> I would fix it later.

Sorry for the force-push and listt so much commits, this issue addressed in [ed75fe7](https://github.com/apache/lucene/pull/12405/commits/ed75fe7a0d67d64e4cfcc0f7e3e1b69326f2ead1) and [9af4ba5](https://github.com/apache/lucene/pull/12405/commits/9af4ba559bb4f128a2a48f38bccb68cbe14ebbfe)

mikemccand commented:
@LuXugang @jpountz it looks like this PR went through some great discussions / iterations and was close towards the end, but it has accumulated some conflicts now?

jpountz commented:
Sorry, since I had approved the PR, I had not understood it was still waiting on me. It's a great change, let's see how to get it in.

jpountz committed:
Merge branch 'main' into skipbydocvalues

jpountz commented:
I did my best at fixing conflicts, @LuXugang are you able to check the changes?

jpountz commented:
Tests fail because the optimization kicks in in more cases than the test expects, it's not clear to me yet if it's a bug or not.

LuXugang commented:
Sure thing  @jpountz , I would work on this  in the next few days.

LuXugang committed:
fix test

LuXugang committed:
Merge branch 'main' into skipbydocvalues

LuXugang commented:
> Tests fail because the optimization kicks in in more cases than the test expects, it's not clear to me yet if it's a bug or not.

You are correct @jpountz , the optimization kicks in and it is not a bug.

In that failed test, after the queue is full with the only one comparator(means Pruning.GREATER_THAN_OR_EQUAL_TO)  , the bottom value is Long.MAX_VALUE, the missing value which is Long.MAX_VALUE would be non-competitive. then optimization kicks in.

jpountz reviewed (approved):
LGTM, thank you!
